msgpack/encode.ts

```diff
@@ -0,0 +1,267 @@
+      const dataView = new DataView(new ArrayBuffer(5));
+      dataView.setUint32(1, len);
+      dataView.setUint8(0, 0xdb);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else {
+      throw new Error(
+        "Cannot safely encode string with size larger than 32 bits",
+      );
+    }
+    byteList.add(encoded);
+    return;
+  }
+
+  if (object instanceof Uint8Array) {
+    if (object.length < EIGHT_BITS) { // bin 8
+      byteList.add(new Uint8Array([0xc4, object.length]));
+    } else if (object.length < SIXTEEN_BITS) { // bin 16
+      const dataView = new DataView(new ArrayBuffer(3));
+      dataView.setUint16(1, object.length);
+      dataView.setUint8(0, 0xc5);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else if (object.length < THIRTY_TWO_BITS) { // bin 32
+      const dataView = new DataView(new ArrayBuffer(5));
+      dataView.setUint32(1, object.length);
+      dataView.setUint8(0, 0xc6);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else {
+      throw new Error(
+        "Cannot safely encode Uint8Array with size larger than 32 bits",
+      );
+    }
+    byteList.add(object);
+    return;
+  }
+
+  if (Array.isArray(object)) {
+    if (object.length < FOUR_BITS) { // fixarray
+      byteList.add(new Uint8Array([0x90 | object.length]));
+    } else if (object.length < SIXTEEN_BITS) { // array 16
+      const dataView = new DataView(new ArrayBuffer(3));
+      dataView.setUint16(1, object.length);
+      dataView.setUint8(0, 0xdc);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else if (object.length < THIRTY_TWO_BITS) { // array 32
+      const dataView = new DataView(new ArrayBuffer(5));
+      dataView.setUint32(1, object.length);
+      dataView.setUint8(0, 0xdd);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else {
+      throw new Error(
+        "Cannot safely encode array with size larger than 32 bits",
+      );
+    }
+
+    for (const obj of object) {
+      encodeSlice(obj, byteList);
+    }
+    return;
+  }
+
+  // If object is a plain object
+  if (object.constructor === Object) {
+    const numKeys = Object.keys(object).length;
+
+    if (numKeys < FOUR_BITS) { // fixarray
+      byteList.add(new Uint8Array([0x80 | numKeys]));
+    } else if (numKeys < SIXTEEN_BITS) { // map 16
+      const dataView = new DataView(new ArrayBuffer(3));
+      dataView.setUint16(1, numKeys);
+      dataView.setUint8(0, 0xde);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else if (numKeys < THIRTY_TWO_BITS) { // map 32
+      const dataView = new DataView(new ArrayBuffer(5));
+      dataView.setUint32(1, numKeys);
+      dataView.setUint8(0, 0xdf);
+      byteList.add(new Uint8Array(dataView.buffer));
+    } else {
+      throw new Error("Cannot safely encode map with size larger than 32 bits");
+    }
+
+    for (const [key, value] of Object.entries(object)) {
+      encodeSlice(key, byteList);
+      encodeSlice(value, byteList);
```

Comment:
**question:** What's the usual choice for undefined values in objects? Right now `{ foo: { optional: undefined, required: 3 }  }` would throw but should it just encode into `{ foo: { required: 3 } }`?

Reply:
Left for a future PR, see discord convo.